src/Compilers/Core/Portable/Serialization/ObjectWriter.cs

```diff
@@ -817,104 +732,6 @@ public void WriteEncoding(Encoding? encoding)
-                    if (writable == null)
-                    {
-                        throw NoSerializationWriterException($"{instance.GetType()} must implement {nameof(IObjectWritable)}");
-                    }
-                }
-
-                var oldDepth = _recursionDepth;
-                _recursionDepth++;
-
-                if (_recursionDepth % MaxRecursionDepth == 0)
-                {
-                    _cancellationToken.ThrowIfCancellationRequested();
-
-                    // If we're recursing too deep, move the work to another thread to do so we
-                    // don't blow the stack.
-                    var task = SerializationThreadPool.RunOnBackgroundThreadAsync(
-                        obj =>
-                        {
-                            WriteObjectWorker((IObjectWritable)obj!);
-                            return null;
-                        },
-                        writable);
-
-                    // We must not proceed until the additional task completes. After returning from a write, the underlying
-                    // stream providing access to raw memory will be closed; if this occurs before the separate thread
-                    // completes its write then an access violation can occur attempting to write to unmapped memory.
-                    //
-                    // CANCELLATION: If cancellation is required, DO NOT attempt to cancel the operation by cancelling this
-                    // wait. Cancellation must only be implemented by modifying 'task' to cancel itself in a timely manner
-                    // so the wait can complete.
-                    task.GetAwaiter().GetResult();
-                }
-                else
-                {
-                    WriteObjectWorker(writable);
-                }
-
-                _recursionDepth--;
-                Debug.Assert(_recursionDepth == oldDepth);
-            }
-        }
-
-        private void WriteObjectWorker(IObjectWritable writable)
-        {
-            _objectReferenceMap.Add(writable, writable.ShouldReuseInSerialization);
-
-            // emit object header up front
-            _writer.Write((byte)TypeCode.Object);
-
-            // Directly write out the type-id for this object.  i.e. no need to write out the 'Type'
-            // tag since we just wrote out the 'Object' tag
-            this.WriteInt32(_binderSnapshot.GetTypeId(writable.GetType()));
-            writable.WriteTo(this);
-        }
-
-        private static Exception NoSerializationTypeException(string typeName)
-        {
-            return new InvalidOperationException(string.Format(Resources.The_type_0_is_not_understood_by_the_serialization_binder, typeName));
-        }
-
-        private static Exception NoSerializationWriterException(string typeName)
-        {
-            return new InvalidOperationException(string.Format(Resources.Cannot_serialize_type_0, typeName));
-        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

I was going to make a comment that the comment below is no longer in compiler and there is a bidirectional map available. But then I noticed that it uses ImmutableDictionarys for it's mapping, which is definitely not perf friendly. As the usage here doesn't look like it needs modifications, would it make sense for there to be a NonMutableBidirectionalMap?

Reply:
yes.  that would make a ton of sense.